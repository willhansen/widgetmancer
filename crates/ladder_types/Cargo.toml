[package]
name = "ladder_types"
version = "0.0.0"
edition.workspace = true

[dependencies]
typenum = "1.17.0"
trybuild = "1.0"
num.workspace = true
derive_more.workspace = true
