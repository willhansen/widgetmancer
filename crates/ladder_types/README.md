A wrapper-type for the phenomenon of subtracting two points resulting in a vector rather than a point.

This crate differs from other approaches (eg euclid) in that rather than the difference of two vectors being a vector, it instead reaches a third level, and so on.

Statically checked

`ThingWithRelativity<Contents, Level=0> ==> T<0>`

`T<0> - T<0> ==> T<1>`
`T<1> - T<1> ==> T<2>`
`T<N> - T<N> ==> T<N+1>`

`T<0> +/- T<1> ==> T<0>`
`T<N> +/- T<N+1> ==> T<0>`

`T<N> + T<M>, where N!=M ==> INVALID`
`T<N> - T<M>, where N!=M and N!=M+1 ==> INVALID`

May have another static parameter for allowing opt-in behavior.  eg allowing `ThingWithRelativity<i32, Level=0>::new(5) + 7`
- Assume no relativity implies level of 0, or whatever the other level is (or one higher?)?
- Maybe an enum for default relativity level?  Options: Same, PlusOne, Zero, Disallow
