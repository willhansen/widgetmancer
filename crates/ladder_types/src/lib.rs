pub mod kiss_ladder;
pub mod typenum_ladder;
pub mod unary_wrapper_ladder;
pub mod numbers_on_ladders;


