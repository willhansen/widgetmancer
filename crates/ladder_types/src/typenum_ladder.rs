use std::ops::Add;
use std::ops::Sub;
use typenum::{U0, U1, U2, U3, U4, B1, Add1, Unsigned};

#[derive(Eq, PartialEq, Debug)]
pub struct LadderRung<T: Unsigned>(std::marker::PhantomData<T>);

impl<T: Unsigned> Default for LadderRung<T> {
    fn default() -> Self {
        Self::new()
    }
}

impl<T: Unsigned> LadderRung<T> {
    pub fn new() -> Self {
        Self(std::marker::PhantomData::<T>)
    }
}

pub type R0 = LadderRung<U0>;
pub type R1 = LadderRung<U1>;
pub type R2 = LadderRung<U2>;
pub type R3 = LadderRung<U3>;
pub type R4 = LadderRung<U4>;
// ...


impl<L> Add<LadderRung<Add1<L>>> for LadderRung<L>
    where
    L: Unsigned + Add<B1>,
    Add1<L>: Unsigned
{
    type Output = LadderRung<L>;

    fn add(self, _rhs: LadderRung<Add1<L>>) -> Self::Output {
        Self::Output::new()
    }
} 

impl<L, R, D> Sub<LadderRung<R>> for LadderRung<L>
    where
    L: Unsigned + SubWithLevelDiff<D>,
    R: Unsigned + Sub<L, Output=D>
{
    type Output = LadderRung<<L as SubWithLevelDiff<D>>::OutputLevel>;

    fn sub(self, _rhs: LadderRung<R>) -> Self::Output {
        Self::Output::new()
    }
} 

pub trait SubWithLevelDiff<D> {
    type OutputLevel: Unsigned;
}

impl<L> SubWithLevelDiff<U0> for L 
where
    L: Unsigned + Add<B1>,
    Add1<L>: Unsigned
{
    type OutputLevel = Add1<L>;
}

impl<L> SubWithLevelDiff<U1> for L
where
    L: Unsigned
{
    type OutputLevel = L;
}



#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_add() {
        assert_eq!(R0::new() + R1::new(), R0::new());
        assert_eq!(R1::new() + R2::new(), R1::new());
    }
    #[test]
    fn test_sub() {
        assert_eq!(R0::new() - R1::new(), R0::new());
        assert_eq!(R1::new() - R2::new(), R1::new());

        assert_eq!(R0::new() - R0::new(), R1::new());
        assert_eq!(R1::new() - R1::new(), R2::new());
    }
}

