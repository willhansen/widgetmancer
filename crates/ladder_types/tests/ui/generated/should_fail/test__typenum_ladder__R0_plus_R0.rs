
use ladder_types::typenum_ladder::*;

fn main() {
   let _ = R0::new() + R0::new();
}
