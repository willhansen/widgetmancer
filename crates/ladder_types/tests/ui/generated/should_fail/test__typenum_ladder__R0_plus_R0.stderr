error[E0308]: mismatched types
 --> tests/ui/generated/should_fail/test__typenum_ladder__R0_plus_R0.rs:5:24
  |
5 |    let _ = R0::new() + R0::new();
  |                        ^^^^^^^^^ expected `LadderRung<UInt<UTerm, B1>>`, found `LadderRung<UTerm>`
  |
  = note: expected struct `ladder_types::typenum_ladder::LadderRung<typenum::uint::UInt<typenum::uint::UTerm, typenum::bit::B1>>`
             found struct `ladder_types::typenum_ladder::LadderRung<typenum::uint::UTerm>`
